src/sentry/integrations/messaging/commands.py

```diff
@@ -74,29 +105,48 @@ def get_all_command_slugs(self) -> Iterable[CommandSlug]:
 class MessagingIntegrationCommandDispatcher(Generic[R], ABC):
     """The set of commands handled by one messaging integration."""
 
+    @property
+    @abstractmethod
+    def integration_spec(self) -> MessagingIntegrationSpec:
+        raise NotImplementedError
+
     @property
     @abstractmethod
     def command_handlers(
         self,
     ) -> Iterable[tuple[MessagingIntegrationCommand, Callable[[CommandInput], R]]]:
         raise NotImplementedError
 
+    def get_event(self, command: MessagingIntegrationCommand) -> MessagingInteractionEvent:
+        return MessagingInteractionEvent(
+            interaction_type=command.interaction_type,
+            provider=self.integration_spec.provider_slug,
+        )
+
     def dispatch(self, cmd_input: CommandInput) -> R:
-        candidate_handlers: list[tuple[CommandSlug, Callable[[CommandInput], R]]] = []
-        for (command, callback) in self.command_handlers:
-            for slug in command.get_all_command_slugs():
-                candidate_handlers.append((slug, callback))
-
-        def parsing_order(handler: tuple[CommandSlug, Callable[[CommandInput], R]]) -> int:
-            # Sort by descending length of arg tokens. If one slug is a prefix of
-            # another (e.g., "link" and "link team"), we must check for the longer
-            # one first.
-            slug, _ = handler
-            return -len(slug.tokens)
-
-        candidate_handlers.sort(key=parsing_order)
-        for (slug, callback) in candidate_handlers:
-            if slug.does_match(cmd_input):
-                arg_input = cmd_input.adjust(slug)
-                return callback(arg_input)
+        @dataclass(frozen=True)
+        class CandidateHandler:
```

Comment:
If you're wondering why I've nested this class definition in such a strange place, it's so that `R` can have the same meaning inside and out. The alternative is kind of ugly: see https://github.com/getsentry/sentry/pull/77820/commits/07b54019715349fa9d555312f2d08ffff5dd9a5f.